src/main/java/emu/grasscutter/game/world/Scene.java

```diff
@@ -484,49 +509,72 @@ public void checkSpawns() {
 			this.broadcastPacket(new PacketSceneEntityDisappearNotify(toRemove, VisionType.VISION_REMOVE));
 		}
 	}
-	
+	public Set<SceneBlock> getPlayerActiveBlocks(Player player){
+		return getScriptManager().getBlocks().stream()
+				.filter(block -> block.contains(player.getPos()))
+				.collect(Collectors.toSet());
+	}
 	public void checkBlocks() {
 		Set<SceneBlock> visible = new HashSet<>();
-		
 		for (Player player : this.getPlayers()) {
-			for (SceneBlock block : getScriptManager().getBlocks()) {
-				if (!block.contains(player.getPos())) {
-					continue;
+			var blocks = getPlayerActiveBlocks(player);
+			visible.addAll(blocks);
+
+			for(var block : blocks){
```

Comment:
This loop for looping through blocks needs to be outside the player loop, otherwise you would be loading multiples of the same blocks if multiple people were in the same area.